doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,322 @@
+      that it shouldn't cause any problems in practice. We can increase
+      the limit later without a breaking API change, but we cannot decrease
+      it later. So we might want to start with something small, like 8.
+
+- We also need to decide what to do for OSes which don't support at least
+  `sendmsg`/`recvmsg`.
+
+  - Don't provide these functions and require all users of these functions to
+    have an alternate code path which doesn't rely on them?
+
+    - Not providing these functions on OSes that don't support
+      at least sendmsg/recvmsg is a simple solution but adds
+      complexity to code using BIO_dgram. (Though it does communicate
+      to code more realistic performance expectations since it
+      knows when these functions are actually available.)
+
+  - Provide these functions and emulate the functionality:
+
+    - However there is a question here as to how we implement
+      the iovec arguments on platforms without `sendmsg`/`recvmsg`. (We cannot
+      use `writev`/`readv` because we need peer address information.) Logically
+      implementing these would then have to be done by copying buffers around
+      internally before calling `sendto`/`recvfrom`, defeating the point of
+      iovecs and providing a performance profile which is surprising to code
+      using BIO_dgram.
+
+    - Another option could be a variable limit on the number of iovecs,
+      which can be queried from BIO_dgram. This would be a constant set
+      when libcrypto is compiled. It would be 1 for platforms not supporting
+      `sendmsg`/`recvmsg`. This again adds burdens on the code using
+      BIO_dgram, but it seems the only way to avoid the surprising performance
+      pitfall of buffer copying to emulate iovec support. There is a fair risk
+      of code being written which accidentially works on one platform but not
+      another, because the author didn't realise the iovec limit is 1 on some
+      platforms. Possibly we could have an “iovec limit” variable in the
+      BIO_dgram which is 1 by default, which can be increased by a call to a
+      function BIO_set_iovec_limit, but not beyond the fixed size discussed
+      above. It would return failure if not possible and this would give client
+      code a clear way to determine if its expectations are met.
+
+Alternate API
+-------------
+
+Could we use a simplified API? For example, could we have an API that returns
+one datagram where BIO_dgram uses `readmmsg` internally and queues the returned
+datagrams, thereby still avoiding extra syscalls but offering a simple API.
+
+The problem here is we want to support “single-copy” (where the data is only
+copied as it is decrypted). Thus BIO_dgram needs to know the final resting place
+of encrypted data at the time it makes the `readmmsg` call.
+
+One option would be to allow the user to set a callback on BIO_dgram it can use
+to request a new buffer, then have an API which returns the buffer:
```

Comment:
This seems like it will interact poorly with applications that really want to use iovecs (so as to read part of the message into a separate header buffer, etc.).  Even letting them allocate their own contiguous buffers still doesn't support that use case.
(It's a clever idea regardless, but I'm not sure it's the best long-term choice.)